Log/Log4perl/Layout/Mozilla.pm

```diff
@@ -0,0 +1,60 @@
+package Log::Log4perl::Layout::Mozilla;
+use 5.10.1;
+use Moo;
+use Sys::Hostname;
+use JSON::MaybeXS ();
+use English qw(-no_match_vars $PID);
+
+use constant LOGGING_FORMAT_VERSION => "2.0";
+
+extends 'Log::Log4perl::Layout';
+
+has 'name' => (
+    is      => 'ro',
+    default => 'Bugzilla',
+);
+
+has 'max_json_length' => (
+    is      => 'ro',
+    isa     => sub { die "must be at least 1024\n" if $_[0] < 1024 },
+    default => 4096,
+);
+
+sub render {
+    my ( $self, $message, $category, $priority, $caller_level ) = @_;
+
+    state $HOSTNAME = hostname();
+    state $JSON     = JSON::MaybeXS->new(
+        indent          => 0,    # to prevent newlines (and save space)
+        ascii           => 1,    # to avoid encoding issues downstream
+        allow_unknown   => 1,    # encode null on bad value (instead of exception)
+        convert_blessed => 1,    # call TO_JSON on blessed ref, if it exists
+        allow_blessed   => 1,    # encode null on blessed ref that can't be converted
+    );
+
+    my $mdc_context = Log::Log4perl::MDC->get_context;
+    my %fields = (
+        message => $message,
+        %$mdc_context,
+    );
+    my %out = (
+        Timestamp => time() * 1e9,
+        Type      => $category,
+        Logger    => $self->name,
+        Hostname  => $HOSTNAME,
+        EnvVersion => LOGGING_FORMAT_VERSION,
+        Severity   => $Log::Log4perl::Level::SYSLOG{$priority},
+        Fields     => \%fields,
+        Pid        => $PID,
+    );
+    my $json_text = $JSON->encode(\%out) . "\n";
+    if (length($json_text) > $self->max_json_length) {
+        %fields = (
+            message => sprintf("msg is larger than %d: %d", $self->max_json_length, length($json_text))
+        );
```

Comment:
it may be better to just change $fields{message} so the MDC context fields are retained.

Reply:
MDC context fields in some cases could be user agents and so on.